Applications/Spire/Include/Spire/Spire/QtFuture.hpp

```diff
@@ -3,36 +3,69 @@
 #include <exception>
 #include <utility>
 #include <Beam/Routines/Async.hpp>
-#include <boost/noncopyable.hpp>
 #include "Spire/Spire/QtPromise.hpp"
-#include "Spire/Spire/Spire.hpp"
 
 namespace Spire {
 
   /*! \brief Encapsulates a QtPromise resolving to a future value.
       \tparam T The type of value to resolve to.
   */
   template<typename T>
-  class QtFuture : private boost::noncopyable {
+  class BaseQtFuture {
     public:
 
       //! The type of value to resolve to.
       using Type = T;
 
+      //! Resolves the QtPromise waiting on this future to an exception.
+      void resolve(std::exception_ptr e);
+
+    protected:
+      Beam::Routines::Eval<Type> m_eval;
+
+      BaseQtFuture(Beam::Routines::Eval<Type> eval);
+      BaseQtFuture(BaseQtFuture&&) = default;
+      BaseQtFuture& operator =(BaseQtFuture&&) = default;
+  };
+
+  /*! \brief Encapsulates a QtPromise resolving to a future value.
+      \tparam T The type of value to resolve to.
+  */
+  template<typename T>
+  class QtFuture : public BaseQtFuture<T> {
+    public:
+
       QtFuture(QtFuture&&) = default;
 
       QtFuture& operator =(QtFuture&&) = default;
 
+      using BaseQtFuture<T>::resolve;
+
       //! Resolves the QtPromise waiting on this future.
       void resolve(Type value);
 
-      //! Resolves the QtPromise waiting on this future to an exception.
-      void resolve(std::exception_ptr e);
+    private:
+      friend struct std::pair<QtFuture, QtPromise<Type>>;
+
+      QtFuture(Beam::Routines::Eval<Type> eval);
```

Comment:
Inherit the constructor.

Reply:
I think there is a VS bug related to this. Inheritance of protected constructors was supposed to be fixed in VS 16.0 (I run 16.4.2), but it seems still broken when used together with `piecewise_construct`. For example, the below code:

```
struct A {
protected:
  A(int x) : m_x(x) {}
private:
  int m_x;
};

struct B: public A {
private:
  using A::A;
  friend struct std::pair<B, int>;
};

std::pair<B, int> factory() {
  return std::pair<B, int>(std::piecewise_construct,
    std::forward_as_tuple(5), std::forward_as_tuple(10));
}

int main() {
  auto b = factory();
  return 0;
}
```

Compiles fine with clang but fails with VS:
` error C2248: '`anonymous-namespace'::A::A': cannot access protected member declared in class '`anonymous-namespace'::A'`

That said, in order to inherit the constructors, I should apparently either make them public or remove piecewise construction.

Reply:
VS sucks. Make ```std::pair``` a friend of ```BaseQtFuture```.

Reply:
Friending with the base class pair, i.e. if I put
`friend struct std::pair<A, int>;`
in the code for A above, does not help.

The only way to make it work is by using CRTP, i.e.:
```
template<typename T>
struct A {
protected:
  A(int x) : m_x(x) {}
private:
  friend struct std::pair<T, int>;
  int m_x;
};

struct B: public A<B> {
private:
  using A::A;
  friend struct std::pair<B, int>;
};

std::pair<B, int> factory() {
  return std::pair<B, int>(std::piecewise_construct,
    std::forward_as_tuple(5), std::forward_as_tuple(10));
}

int main() {
  auto b = factory();
  return 0;
}
```

I can do that for `QtFuture` if it is not an overkill.